### DISCUSSION TITLE
Coverage testing an API?

### DISCUSSION THREAD

**Discussion by rowanwins** [vitest-dev/vitest#5473] [Q&A] [answered]
Hi there,

I have a set of Vitest tests which test API responses from an ExpressJS app that I'm running. My tests run by sending a fetch request to a set of ExpressJS endpoint, eg
````
const ENDPOINT = 'http://localhost:3000/widgets'

test("Get Widgets - Unauthenticated", async () => {
  const response = await getWithoutCredentials(ENDPOINT)
  expect(response.status).toBe(401)
  expect(response.ok).toBe(false)
})

test("Get Widgets - Authenticated", async () => {
  const response = await getWithCredentials(ENDPOINT)
  expect(response.status).toBe(200)
  const data = await response.json()
  expect(data.length).toBe(2)
})
````

If I run the tests everything works as expected for checking the API response. 

However if I try and run a coverage check against my codebase I'm not able to generate any sensible outputs. So in the case above I'd expect to see some coverage for `src/routes/widgets.ts` and `src/controllers/widgets.ts`.
- If I set `istanbul` as my coverage provider then the coverage check fails, it seems unhappy with something in the typescript
- If I set the coverage provider to `v8` then then coverage check completes however files which I'd expect to have some coverage instead have none.

I'm suspect that this is because my tests are simply sending fetch requests off to my API, and so the coverage testing doesn't know to look for action within the `/src/routes/widgets.ts` and `src/controllers/widgets.ts` files.

Has anyone attempted something similar with Vite and have any suggestions? 

Thanks,
Rowan

**Comment by AriPerkkio** [comment]
How is the API started? Could you set up an example project for this?

**Reply by rowanwins** [reply]
Hi @AriPerkkio 

I've put together a minimal repro here
https://github.com/rowanwins/express-api-vitest-coverage

Of note
- There is a `globalSetup` file which starts up the expressjs server
-  Running `npm run test` runs some tests within `/tests/api`
   -  these make fetch requests to check the API response
- Running `npm run test:coverage` ideally should look across my `/src` directory and see which parts of the code are being tested by my requests

**Reply by hi-ogawa** [reply, marked as answer]
`globalSetup` runs in a main process while [`setupFiles`](https://vitest.dev/config/#setupfiles) runs in a same runtime as your test code, which is where coverage is taken. So you may try `setupFiles`, but its setup/teardown runs for each test file, so you might need to adjust your test structure.

Or as you noticed you could use `supertest`, which is a wrapper to launch server (inside test code) and write an request/response assertion in an opinionated manner, but I think what happens at lower level isn't so different from your style of running `app.listen` manually and writing assertion using `fetch` directly.

**Reply by rowanwins** [reply]
Ahh that is very helpful thanks @hi-ogawa - yep that seems to be the crux of my problem, if I run my server startup in  in a setupFile then indeed the coverage check works 🎉

**Comment by rowanwins** [comment]
Maybe another option is to look at [`supertest`](https://github.com/ladjs/supertest) for doing my request/response as it hooks into the express app rather than just sending fetch requests....

There is a [blog post](https://simonkkaranja.medium.com/node-testing-jest-supertest-343c9b650d89) which outlines this sort of approach but with jest (instead of Vitest) and supertest.
